File: app/src/test/java/com/mytaxiapp/VehicleInterfaceTest.java
package com.mytaxiapp;

import android.app.Application;

import com.mytaxiapp.data.model.Placemark;
import com.mytaxiapp.data.model.Vehicles;
import com.mytaxiapp.data.remote.VehicleInteractor;
import com.mytaxiapp.data.remote.VehicleInterface;
import com.mytaxiapp.ui.listvehicles.VehiclePresenter;
import com.mytaxiapp.ui.listvehicles.VehicleView;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.observers.TestSubscriber;

import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.when;

/**
 * @author Tosin Onikute.
 */

public class VehicleInterfaceTest {

    @Mock
    VehicleInterface vehicleInterface;

    @Mock
    Application application;

    @Mock
    VehicleInteractor vehicleInteractor;

    @Mock
    VehicleView vehicleView;

    private VehiclePresenter presenter;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        presenter = new VehiclePresenter(application, vehicleInteractor);
        presenter.attachView(vehicleView);
    }

    @Test
    public void testAPIResponse() throws Exception {
        // for this test, we use the first placemark
        String name = "HH-GO8522";
        String address = "Lesserstraße 170, 22049 Hamburg";
        String vin = "WME4513341K565439";

        // when
        when(vehicleInterface.getVehicles()).thenReturn(Observable.just(vehiclesList()));

        // given
        given(vehicleInterface.getVehicles()).willReturn(Observable.just(vehiclesList()));

        // When
        TestSubscriber<Vehicles> subscriber = new TestSubscriber<>();
        vehicleInterface.getVehicles().subscribe(subscriber);

        // Then
        subscriber.awaitTerminalEvent();
        subscriber.assertNoErrors();

        List<Vehicles> onNextEvents = subscriber.getOnNextEvents();
        List<Placemark> placemarkInfo = onNextEvents.get(0).getPlacemarks();

        // assert for name
        Assert.assertEquals(name, placemarkInfo.get(0).getName());

        // assert for address
        Assert.assertEquals(address, placemarkInfo.get(0).getAddress());

        // assert for vin
        Assert.assertEquals(vin, placemarkInfo.get(0).getVin());

    }

    private Vehicles vehiclesList() {

        // Setting up the values for test
        Vehicles vehicles = new Vehicles();

        List<Placemark> listPlacemarks = new ArrayList<Placemark>();

        List<Double> coordinates = new ArrayList<Double>();
        coordinates.add(10.07526);
        coordinates.add(53.59301);
        coordinates.add(0.0);

        Placemark placemark = new Placemark();
        placemark.setName("HH-GO8522");
        placemark.setAddress("Lesserstraße 170, 22049 Hamburg");
        placemark.setCoordinates(coordinates);
        placemark.setEngineType("CE");
        placemark.setExterior("UNACCEPTABLE");
        placemark.setFuel(42);
        placemark.setInterior("UNACCEPTABLE");
        placemark.setVin("WME4513341K565439");
        listPlacemarks.add(placemark);

        // add all the placemark data into Vehicle data structure
        vehicles.setPlacemarks(listPlacemarks);

        return vehicles;
    }

}

File: app/src/test/java/com/mytaxiapp/VehicleInterfaceTest.kt
package com.mytaxiapp

import android.app.Application

import com.mytaxiapp.data.model.Placemark
import com.mytaxiapp.data.model.Vehicles
import com.mytaxiapp.data.remote.VehicleInteractor
import com.mytaxiapp.data.remote.VehicleInterface
import com.mytaxiapp.ui.listvehicles.VehiclePresenter
import com.mytaxiapp.ui.listvehicles.VehicleView

import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations

import java.util.ArrayList

import rx.Observable
import rx.observers.TestSubscriber

import org.mockito.BDDMockito.given
import org.mockito.Mockito.`when`

/**
 * @author Tosin Onikute.
 */

class VehicleInterfaceTest {

    @Mock
    internal var vehicleInterface: VehicleInterface? = null

    @Mock
    internal var application: Application? = null

    @Mock
    internal var vehicleInteractor: VehicleInteractor? = null

    @Mock
    internal var vehicleView: VehicleView? = null

    private var presenter: VehiclePresenter? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        presenter = VehiclePresenter(application!!, vehicleInteractor!!)
        presenter!!.attachView(vehicleView!!)
    }

    @Test
    @Throws(Exception::class)
    fun testAPIResponse() {
        // for this test, we use the first placemark
        val name = "HH-GO8522"
        val address = "Lesserstraße 170, 22049 Hamburg"
        val vin = "WME4513341K565439"

        // when
        `when`(vehicleInterface!!.vehicles).thenReturn(Observable.just(vehiclesList()))

        // given
        given(vehicleInterface!!.vehicles).willReturn(Observable.just(vehiclesList()))

        // When
        val subscriber = TestSubscriber<Vehicles>()
        vehicleInterface!!.vehicles.subscribe(subscriber)

        // Then
        subscriber.awaitTerminalEvent()
        subscriber.assertNoErrors()

        val onNextEvents = subscriber.onNextEvents
        val placemarkInfo = onNextEvents[0].placemarks

        // assert for name
        Assert.assertEquals(name, placemarkInfo!![0].name)

        // assert for address
        Assert.assertEquals(address, placemarkInfo[0].address)

        // assert for vin
        Assert.assertEquals(vin, placemarkInfo[0].vin)

    }

    private fun vehiclesList(): Vehicles {

        // Setting up the values for test
        val vehicles = Vehicles()

        val listPlacemarks = ArrayList<Placemark>()

        val coordinates = ArrayList<Double>()
        coordinates.add(10.07526)
        coordinates.add(53.59301)
        coordinates.add(0.0)

        val placemark = Placemark()
        placemark.name = "HH-GO8522"
        placemark.address = "Lesserstraße 170, 22049 Hamburg"
        placemark.coordinates = coordinates
        placemark.engineType = "CE"
        placemark.exterior = "UNACCEPTABLE"
        placemark.fuel = 42
        placemark.interior = "UNACCEPTABLE"
        placemark.vin = "WME4513341K565439"
        listPlacemarks.add(placemark)

        // add all the placemark data into Vehicle data structure
        vehicles.placemarks = listPlacemarks

        return vehicles
    }

}
